AP_Proximity: add RPLidar S2 support

tatsuy opened the pull request:
### Summary
Add RPLidar S2 support to AP_Proximity.

This is a follow-up PR for #26835  which included the following changes:
- RPLidar S2 support (originally contributed by @mirkix)
- ~Removed memory copy for each measurement (by @mirkix)~
- Added Dense Express (capsulated) scan support for RPLidar S2 with checksum validation to reject corrupted packets
- ~Added a SITL model for RPLidar S2~ move to #31730
- ~Extended the serial corruption simulator~ move to #31752
- Added a ProximitySensors autotest entry for rplidars2

### Classification & Testing (check all that apply and add your own)

- [x] Checked by a human programmer
- [ ] Non-functional change
- [ ] No-binary change
- [ ] Infrastructure change (e.g. unit tests, helper scripts)
- [x] Automated test(s) verify changes (e.g. unit test, autotest)
- [x] Tested manually, description below (e.g. SITL)
- [x] Tested on hardware
- [ ] Logs attached
- [x] Logs available on request

### Description
Reason for this follow-up PR
I tested #26835 and confirmed that valid distance data could be obtained.
But, I noticed that the reported distance values sometimes became abnormal during operation, and once this happened, the driver could continue outputting random values afterward.
After investigating the issue, I found that noise or intermittent contact on the RPLidar S2 TX line could corrupt serial frames, causing the driver to lose synchronization with scan packets.
To address this problem, this PR adds checksum validation for S2 Dense Express packets so that corrupted packets are discarded.
This helps prevent invalid data from being accepted when noise is present and reduces persistent misinterpretation caused by loss of synchronization.

### Settings
PRXx_TYPE 5 (RPLidarA2)
SERIALx_PROTOCOL 11 (Lidar360)
SERIALx_BAUD 1000

### What I tested
- SITL test
Confirmed S2 (and A1/A2) work correctly in SITL:
`sim_vehicle.py --gdb --debug -v ArduCopter -A --serial5=sim:rplidars2 --speedup=1 -l 51.8752066,14.6487840,0,0 --map`
`autotest.py --speedup=20 build.Copter test.Copter.ProximitySensors`

- Serial Corruption Testing (using #31752)
set `AP_SIM_SERIALDEVICE_CORRUPTION_ENABLED` to 1
run SITL under Valgrind (-V) and confirmed no memory errors even under heavy corruption. (confirmed S2/A2)
`sim_vehicle.py --gdb --debug -v ArduCopter -A --serial5=sim:rplidars2 --speedup=1 -l 51.8752066,14.6487840,0,0 -V`

- Bench testing on CubeOrangePlus
Tested with CubeOrangePlus + RPLIDAR S2/A2
Tested with Rover at 50 Hz loop rate using S2/A2

- HW Testing
I manually wiggled the TX line on an RPLidar S2 unit to inject real-world serial corruption.
  Confirmed that:
  - corrupted packets are discarded via checksum validation,
  - the driver stays synchronised,
  - descriptor parsing no longer gets stuck in repeated RESETs,
  - only valid distance data continues to be used even while corruption is being injected.

Messages captured with `RP_DEBUG_LEVEL=1` during HW testing (CubeOrangePlus, RPLIDAR S2):
```
2026/04/21 17:25:28 : EXPRESS/DENSE checksum error
2026/04/21 17:25:28 : EXPRESS/DENSE checksum error
2026/04/21 17:25:28 : EXPRESS: no header candidate, dropping 327 bytes
2026/04/21 17:25:28 : EXPRESS: no header candidate, dropping 302 bytes
<--->
2026/04/21 17:24:34 : LIDAR NO DATA
2026/04/21 17:24:34 : LIDAR NO DATA
2026/04/21 17:24:34 : Sent EXPRESS (Dense) scan request
2026/04/21 17:24:34 : RPLidar S2 hw=18 fw=2.1
2026/04/21 17:24:34 : Received DEVICE_INFO
```
I confirmed the error detection and resynchronization by manually wiggled the TX line
and continuasly obtain the normal value from S2.

- Flight testing
 Flight-tested the same functionality on an actual copter with v4.6:
 https://github.com/tatsuy/ardupilot/commits/RPLidarS2-4.6/
 It worked as expected

I only have A2 and S2 hardware, so I could not test A1, C1, or S1 directly.
The new code path is enabled only for S2 via _use_dense_express; all other models continue to use the existing path unchanged.
I also confirmed on A2 that existing hardware still works correctly with this change.

mirkix committed:
AP_Proximity: add RPLidar S2

tatsuy committed:
Tools: add RPLidar S2 to ProximitySensors autotest

peterbarker reviewed (changes requested):
Why did this require rewriting the entire parser?  I can't see anything in your comments or the code which made this required.

Please remove the parser rewrite if it's not absolutely required.  If it is then create a PR which changes it.  A parser rewrite will require testing on each of the supported RPLidars - the simulators are good but not perfect.

Partial review only.

tridge commented on libraries/AP_Proximity/AP_Proximity_RPLidarA2.h:
```diff
@@ -70,27 +70,46 @@ class AP_Proximity_RPLidarA2 : public AP_Proximity_Backend_Serial
     uint8_t _sync_error;
     uint16_t _byte_count;
 
+    uint8_t _express_stream[EXPRESS_STREAM_BUFFER_SIZE];
```
thats a lot of memory. Is this needed for all RPLidar varients? This could impact lower memory boards like F405

tpwrules replied:
If you are parsing in 84-byte chunks why do you need an 8192 byte buffer? Maybe 84*4?

tatsuy replied:
@tridge @tpwrules 

Thanks, I agree that 8KB is a lot of memory.

I tested this with real RPLidar S2 hardware. On Copter, a small buffer would be enough: read_max was about 336–420 bytes and no overflow was observed. The issue is Rover, where AP_Proximity is currently scheduled at around 50Hz, so one update can accumulate about 1.3–1.5KB of Dense Express data.

In the Rover 50Hz test, read_max reached 1492 bytes and stream_max reached about 7.2KB, with stream overflow observed. So 84 * 4 is not enough for Rover with the current scheduling.

The 8KB buffer was chosen as a compromise to keep S2 Dense Express usable on Rover as well.

Ideally, Rover could avoid this by running the main loop at 400Hz and updating AP_Proximity at 200Hz, similar to Copter. With that scheduler change, a much smaller buffer would likely be enough. But with the current Rover 50Hz scheduling, the S2 Dense Express path needs a larger stream buffer to remain usable.
https://github.com/ArduPilot/ardupilot/blob/master/Rover/Rover.cpp#L87

rmackay9 replied:
Hi @tatsuy,

Yes, we've seen this issue with Rover/Boat quite a few times in the forums and I've always just recommended that users update the main loop to 200hz (or 400hz).  This led us to add a note explaining this to the setup wiki page: https://ardupilot.org/copter/docs/common-rplidar-a2.html

tatsuy replied:
@rmackay9 
Thanks, that makes sense. I saw the wiki note as well.

But, changing SCHED_LOOP_RATE alone may not be enough for Rover/Boat. In Rover, AP_Proximity is also explicitly scheduled in scheduler_tasks at 50Hz, so the proximity backend can still run at 50Hz even if the main loop is increased.

So I think users may also need to change the AP_Proximity task rate, for example to 200Hz like Copter:
```SCHED_TASK_CLASS(AP_Proximity,`        &rover.g2.proximity,     update,         50,  200,  27),```

rmackay9 replied:
Hi @tatsuy,

Yes, changing the Rover scheduler to run the proximity at 200hz sounds like a good idea.  we seem to have many items in the Rover scheduler that run faster than Rover's default main loop rate.  This change doesn't need to be part of this PR though unless you want it to be.  up to you

tatsuy replied:
@rmackay9 Thank you!
I've created the PR #33033 to run the proximity at 200hz.

tridge reviewed (commented):
claude also gives this: 

AP_Proximity_RPLidarA2.cpp:188 - Lost angular resolution: parse_response_express() collapses all 40 cabin distances in a Dense block to a single min_distance_m and pushes only one face update per block. Each cabin spans ~0.5° of arc → 40 cabins ≈ 20° of azimuth, but the code attributes all of them to the block's single start_angle_deg_raw. The existing A2 path uses database_push(angle, dist) per sample. Sweep cab_angle = start + i * angle_step and push each cabin individually.

tatsuy committed:
AP_Proximity: add RPLidar S2 Dense Express scan

- Send EXPRESS_SCAN command (Dense capsule mode) for S2 only
- Per-packet checksum validation (sync nibbles + XOR)
- 672-byte stream buffer and read cap (8 Dense blocks each, sized for
  200 Hz AP_Proximity update) with overflow and underflow protection
- 1 ms time guard to prevent blocking the main loop
- Direct processing in get_readings() to bypass the generic UART
  buffer path
- Each Dense block (40 cabins) is pushed once at the angle of its
  shortest valid cabin (start + min_k * 0.1125 deg); ignore_reading()
  is applied at the winner, with a per-cabin rescan fallback when the
  winner lands inside an ignore zone
- Existing devices (A1/A2/A2M12/C1/S1) are unaffected:
  _use_dense_express is false for all non-S2 models
- Add AP_PROXIMITY_RPLIDAR_EXPRESSSCAN_ENABLED

tatsuy commented:
> claude also gives this:
> 
> AP_Proximity_RPLidarA2.cpp:188 - Lost angular resolution: parse_response_express() collapses all 40 cabin distances in a Dense block to a single min_distance_m and pushes only one face update per block. Each cabin spans ~0.5° of arc → 40 cabins ≈ 20° of azimuth, but the code attributes all of them to the block's single start_angle_deg_raw. The existing A2 path uses database_push(angle, dist) per sample. Sweep cab_angle = start + i * angle_step and push each cabin individually.

@tridge 
Thanks. The previous implementation could introduce up to about 4.39 deg of angular error because it selected the shortest distance from the 40 cabins but attributed it to the block start_angle.

I fixed that by computing the angle from the selected cabin index instead of using start_angle. It still pushes only one closest valid sample per Dense block, but the pushed sample now uses its corrected per-cabin angle.

I also tested a local prototype that pushes all 40 cabins individually, but it caused a significant CPU load increase on CubeOrangePlus. So I think the current approach is a better trade-off between angular accuracy and CPU cost.

tatsuy commented:
I also checked AP_Proximity::update performance (tasks.txt) at 200Hz.

AP_Proximity::update (200Hz)

| Sensor / mode                         | MIN | MAX | AVG | OVR | SLP | TOT   |
|---------------------------------------|----:|----:|----:|----:|----:|------:|
| RPLIDAR A2 (non-Express scan)         |   3 | 112 |  43 |   0 |   0 |  3.6% |
| RPLIDAR S2 (Express scan - this PR)   |  42 | 144 |  79 |   0 |   0 |  1.7% |
| RPLIDAR S2 (full per-cabin push)      | 141 | 549 | 327 | 999 |   0 | 22.9% |

The implementation in this PR had no scheduler overruns on CubeOrangePlus, while the full per-cabin push prototype was much heavier. For that reason I avoided the full per-cabin approach and kept this PR to one corrected closest sample per Dense block.

peterbarker commented on libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp:
```diff
@@ -425,4 +494,288 @@ void AP_Proximity_RPLidarA2::parse_response_health()
+        return (AP_HAL::micros() - start_us) > MAX_US;
+    };
+
+    auto ensure_space = [&](uint16_t need) {
```
PR I created recently attempted to use lambdas and the project is not willing to use them.  Sorry, you'll need to remove these.

tatsuy replied:
I removed lambdas

peterbarker commented on libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp:
```diff
@@ -425,4 +494,288 @@ void AP_Proximity_RPLidarA2::parse_response_health()
     Debug(1, "LIDAR Healthy");
 }
 
+// verify Dense capsulated (Express) block checksum
```
I think this and the memory are enough to warrant a separate `#if AP_PROXYIMITY_RPLIDAR_EXPRESSCAN_ENABLED`

tatsuy replied:
I added AP_PROXIMITY_RPLIDAR_EXPRESSSCAN_ENABLED

peterbarker reviewed (dismissed):
Sorry, the lambda thing won't be a small change but it's not viable for the project at the moment.

tatsuy commented:
@peterbarker Thanks for the review.
I removed lambdas, added AP_PROXIMITY_RPLIDAR_EXPRESSSCAN_ENABLED.

The diff is here:
https://github.com/tatsuy/ardupilot/commit/bf40f77af0aee5f4962b7e8b1d565034dc4c6658

I also tested with CubeOrangePlus + RPLIDAR S2.
